crates/qasm3/src/exporter.rs

```diff
@@ -0,0 +1,1412 @@
+        );
+        Ok(identifier)
+    }
+
+    fn push_scope(&mut self) {
+        self.symbols.push(HashMap::new());
+        self.bitinfo.push(HashMap::new());
+        self.reginfo.push(HashMap::new());
+    }
+
+    fn pop_scope(&mut self) {
+        self.symbols.pop();
+        self.bitinfo.pop();
+        self.reginfo.pop();
+    }
+
+    fn new_context(&mut self) -> Self {
+        let mut new_table = SymbolTable::new();
+        new_table.gates.clone_from(&self.gates);
+        new_table.stdgates.clone_from(&self.stdgates);
+        new_table
+    }
+}
+
+pub struct Exporter {
+    includes: Vec<String>,
+    basis_gates: Vec<String>,
+    disable_constants: bool,
+    allow_aliasing: bool,
+    indent: String,
+}
+
+impl Exporter {
+    pub fn new(
+        includes: Vec<String>,
+        basis_gates: Vec<String>,
+        disable_constants: bool,
+        allow_aliasing: bool,
+        indent: String,
+    ) -> Self {
+        Self {
+            includes,
+            basis_gates,
+            disable_constants,
+            allow_aliasing,
+            indent,
+        }
+    }
+
+    pub fn dumps(&self, circuit_data: &CircuitData, islayout: bool) -> ExporterResult<String> {
+        let mut builder = QASM3Builder::new(
+            circuit_data,
+            islayout,
+            self.includes.clone(),
+            self.basis_gates.clone(),
+            self.disable_constants,
+            self.allow_aliasing,
+        );
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                BasicPrinter::new(&mut output, self.indent.to_string(), false)
+                    .visit(&Node::Program(&program));
+                Ok(output)
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
+
+    pub fn dump<W: Write>(
+        &self,
+        circuit_data: &CircuitData,
+        islayout: bool,
+        writer: &mut W,
+    ) -> ExporterResult<()> {
+        let mut builder = QASM3Builder::new(
+            circuit_data,
+            islayout,
+            self.includes.clone(),
+            self.basis_gates.clone(),
+            self.disable_constants,
+            self.allow_aliasing,
+        );
+
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                let mut printer = BasicPrinter::new(&mut output, self.indent.to_string(), false);
+                printer.visit(&Node::Program(&program));
+                drop(printer);
+                let _ = writer.write_all(output.as_bytes());
+                Ok(())
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
+}
+
+pub struct QASM3Builder {
+    _builtin_instr: HashSet<&'static str>,
+    loose_bit_prefix: &'static str,
+    loose_qubit_prefix: &'static str,
+    _gate_param_prefix: &'static str,
+    _gate_qubit_prefix: &'static str,
```

Comment:
If these are necessarily `&'static str`, we most likely don't need to store references to them, but can just pull them from the outer scope.